# Day3 - PRACTICAL

##FUMA   

In order to learn more about which variants are likely to have functional impact on a trait, we will have to do post-GWAS functional association of all associated SNPs. We need to know where our significant SNPs are located, what kind of variants they are (intergenic, intronic, exonic), if exonic; what kind (missense, nonsense etc.), what SNPs they are in LD with, if they are known to influence expression of a gene, etc. etc. This information is available from many different databases. Today we will work with some of these databases.     

[Part 1: Functional annotation without FUMA](FUMApractical2023_PART1)

[Part 2: Functional annotation without FUMA](FUMApractical2023_PART2)

FUMA GWAS uses information from publicly available annotation and mapping datasets (see https://fuma.ctglab.nl/tutorial#celltype â€“ Data sets)  
