# Day3 - PRACTICAL

## Functional Mapping and Annotation of Genome-Wide Association Studies (FUMA) 

### FOCUS and LEARNING GOALS
> The aim for this session is to get familiar with platforms to annotate,  
> prioritize, visualize and interpret GWAS results.
> Annotating the known function of SNPs and genes in genomic areas of interest
> may bring us closer to a new understanding of biology. The take home message 
> from this session is that functional mapping and annotation may 
> help to provide an informative interpretion our findings.

In order to learn more about which variants are likely to have functional impact on a trait, we will have to do post-GWAS functional association of all associated SNPs. We need to know where our significant SNPs are located, what kind of variants they are (intergenic, intronic, exonic), if exonic; what kind (missense, nonsense etc.), what SNPs they are in LD with, if they are known to influence expression of a gene, etc. etc. This information is available from many different databases. Today we will work with some of these databases.     

This practical is in two parts:    

Firstly, we will follow-up the a findings manually.     

[Part 1: Functional annotation without FUMA](FUMApractical2023_PART1)

Secondly, we will use the FUMA tool.  

[Part 2: Functional annotation without FUMA](FUMApractical2023_PART2)

FUMA GWAS uses information from publicly available annotation and mapping datasets (see https://fuma.ctglab.nl/tutorial#celltype â€“ Data sets)  
